vllm/engine/async_llm_engine.py

```diff
@@ -278,15 +324,33 @@ async def step_async(
                 virtual_engine=virtual_engine,
                 num_lookahead_slots=scheduler_outputs.num_lookahead_slots,
                 running_queue_size=scheduler_outputs.running_queue_size,
-                finished_requests_ids=finished_requests_ids)
+                finished_requests_ids=finished_requests_ids,
+                last_sampled_token_ids=last_sampled_token_ids)
+            # Execute the model.
             output = await self.model_executor.execute_model_async(
                 execute_model_req)
+            # we need to do this here so that last step's sampled_token_ids can
+            # be passed to the next iteration for PP.
+            if self.scheduler_config.is_multi_step:
+                self._cache_output_for_multi_step(virtual_engine, output)
```

Comment:
Ditto. We could probably name it `_update_cached_scaheduler_output`.

Reply:
[01500e4](https://github.com/vllm-project/vllm/pull/7000/commits/01500e42b8f7aada12a0f9a77ca0c6cc727a6443)